Bug: Parsing errors in nested selectors

### Environment

ESLint version: `9.26.0`
@eslint/css version: `0.7.0`
Node version: `23.6.1`
npm version: `10.9.0`
Operating System: `Debian 12`

### Which language are you using?

stylesheet

### What did you do?

<details>
<summary>Configuration</summary>

```js
import css from "@eslint/css";
import {defineConfig} from "eslint/config";

export default defineConfig([
  {files: ["**/*.css"], plugins: {css}, language: "css/css", extends: ["css/recommended"]},
]);

```
</details>

I get parsing errors in nested selectors - see comments in the following CSS example:

```css
.class1 {
  background-color: blue;

  span { /* <------------- Parsing error: Colon is expected */
    background-color: green;
  }
}

.class2 {
  background-color: yellow;

  div[class^="x"] { /* <------------- Parsing error: Colon is expected */
    background-color: red;
  }
}

.class3 {
  background-color: black;

  > thead { /* <------------- Parsing error: Identifier is expected */
    background-color: white;
  }
}
```


### What did you expect to happen?

No errors.

### What actually happened?

```bash
   4:8  error  Parsing error: Colon is expected
  12:6  error  Parsing error: Colon is expected
  20:3  error  Parsing error: Identifier is expected
```

### Link to Minimal Reproducible Example

-

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

-

--- a/README.md
+++ b/README.md
@@ -60,6 +60,7 @@ export default [
 | :----------------------------------------------------------------------- | :------------------------------------ | :-------------: |
 | [`no-duplicate-imports`](./docs/rules/no-duplicate-imports.md)           | Disallow duplicate @import rules      |       yes       |
 | [`no-empty-blocks`](./docs/rules/no-empty-blocks.md)                     | Disallow empty blocks                 |       yes       |
+| [`no-important`](./docs/rules/no-important.md)                           | Disallow !important flags             |       yes       |
 | [`no-invalid-at-rules`](./docs/rules/no-invalid-at-rules.md)             | Disallow invalid at-rules             |       yes       |
 | [`no-invalid-properties`](./docs/rules/no-invalid-properties.md)         | Disallow invalid properties           |       yes       |
 | [`prefer-logical-properties`](./docs/rules/prefer-logical-properties.md) | Enforce the use of logical properties |       no        |

--- a/docs/rules/no-important.md
+++ b/docs/rules/no-important.md
@@ -1,6 +1,6 @@
 # no-important
 
-Disallow !important flags.
+Disallow `!important` flags.
 
 ## Background
 
@@ -33,7 +33,7 @@ Examples of correct code:
 
 ## When Not to Use It
 
-You may disable this rule in these specific cases:
+You may disable this rule if you are using `!important` in these specific cases:
 
 - Overriding third-party styles where you lack control over the source CSS.
 - Fixing accessibility issues (e.g., enforcing focus states or color contrast).

--- a/docs/rules/no-important.md
+++ b/docs/rules/no-important.md
@@ -1,0 +1,45 @@
+# no-important
+
+Disallow !important flags.
+
+## Background
+
+The `!important` flag is a CSS declaration modifier that overrides normal cascade behavior, forcing a declaration to take precedence over competing rules—regardless of specificity or source order. While it can be necessary in rare cases (e.g., overriding third-party styles or enforcing accessibility fixes), it’s widely considered an anti-pattern because:
+
+- It breaks the natural cascade of CSS
+- It makes debugging more difficult
+- It can lead to specificity wars where developers keep adding more `!important` declarations to override each other
+- It makes the code harder to maintain
+
+## Rule Details
+
+This rule warns when it detects the `!important` flag in declarations.
+
+Examples of incorrect code:
+
+```css
+.foo {
+	color: red !important;
+}
+```
+
+Examples of correct code:
+
+```css
+.foo {
+	color: red;
+}
+```
+
+## When Not to Use It
+
+You may disable this rule in these specific cases:
+
+- Overriding third-party styles where you lack control over the source CSS.
+- Fixing accessibility issues (e.g., enforcing focus states or color contrast).
+- Working with legacy code where refactoring isn’t feasible.
+
+## Prior Art
+
+- [`important`](https://github.com/CSSLint/csslint/wiki/Disallow-%21important)
+- [`declaration-no-important`](https://stylelint.io/user-guide/rules/declaration-no-important/)

--- a/docs/rules/no-important.md
+++ b/docs/rules/no-important.md
@@ -35,9 +35,9 @@ Examples of correct code:
 
 You may disable this rule if you are using `!important` in these specific cases:
 
-- Overriding third-party styles where you lack control over the source CSS.
-- Fixing accessibility issues (e.g., enforcing focus states or color contrast).
-- Working with legacy code where refactoring isn’t feasible.
+- Overriding third-party styles where you lack control over the source CSS
+- Fixing accessibility issues (e.g., enforcing focus states or color contrast)
+- Working with legacy code where refactoring isn’t feasible
 
 ## Prior Art
 

--- a/src/index.js
+++ b/src/index.js
@@ -11,6 +11,7 @@ import { CSSLanguage } from "./languages/css-language.js";
 import { CSSSourceCode } from "./languages/css-source-code.js";
 import noEmptyBlocks from "./rules/no-empty-blocks.js";
 import noDuplicateImports from "./rules/no-duplicate-imports.js";
+import noImportant from "./rules/no-important.js";
 import noInvalidProperties from "./rules/no-invalid-properties.js";
 import noInvalidAtRules from "./rules/no-invalid-at-rules.js";
 import preferLogicalProperties from "./rules/prefer-logical-properties.js";
@@ -32,6 +33,7 @@ const plugin = {
 	rules: {
 		"no-empty-blocks": noEmptyBlocks,
 		"no-duplicate-imports": noDuplicateImports,
+		"no-important": noImportant,
 		"no-invalid-at-rules": noInvalidAtRules,
 		"no-invalid-properties": noInvalidProperties,
 		"prefer-logical-properties": preferLogicalProperties,
@@ -44,6 +46,7 @@ const plugin = {
 			rules: /** @type {const} */ ({
 				"css/no-empty-blocks": "error",
 				"css/no-duplicate-imports": "error",
+				"css/no-important": "error",
 				"css/no-invalid-at-rules": "error",
 				"css/no-invalid-properties": "error",
 				"css/use-baseline": "warn",

--- a/src/rules/no-important.js
+++ b/src/rules/no-important.js
@@ -1,0 +1,91 @@
+/**
+ * @fileoverview Rule to disallow !important flags.
+ */
+
+//-----------------------------------------------------------------------------
+// Imports
+//-----------------------------------------------------------------------------
+
+import { findOffsets } from "../util.js";
+
+//-----------------------------------------------------------------------------
+// Type Definitions
+//-----------------------------------------------------------------------------
+
+/**
+ * @import { CSSRuleDefinition } from "../types.js"
+ * @typedef {"unexpectedImportant"} NoImportantMessageIds
+ * @typedef {CSSRuleDefinition<{ RuleOptions: [], MessageIds: NoImportantMessageIds }>} NoImportantRuleDefinition
+ */
+
+//-----------------------------------------------------------------------------
+// Rule Definition
+//-----------------------------------------------------------------------------
+
+/** @type {NoImportantRuleDefinition} */
+export default {
+	meta: {
+		type: "problem",
+
+		docs: {
+			description: "Disallow !important flags",
+			recommended: true,
+			url: "https://github.com/eslint/css/blob/main/docs/rules/no-important.md",
+		},
+
+		messages: {
+			unexpectedImportant: "Unexpected !important flag found.",
+		},
+	},
+
+	create(context) {
+		return {
+			Declaration(node) {
+				if (node.important) {
+					const declarationText = context.sourceCode.getText(node);
+					const importantPattern = /!(\s|\/\*.*?\*\/)*important/iu;
+					const importantMatch =
+						importantPattern.exec(declarationText);
+
+					const {
+						lineOffset: startLineOffset,
+						columnOffset: startColumnOffset,
+					} = findOffsets(declarationText, importantMatch.index);
+
+					const {
+						lineOffset: endLineOffset,
+						columnOffset: endColumnOffset,
+					} = findOffsets(
+						declarationText,
+						importantMatch.index + importantMatch[0].length,
+					);
+
+					const nodeStartLine = node.loc.start.line;
+					const nodeStartColumn = node.loc.start.column;
+					const startLine = nodeStartLine + startLineOffset;
+					const endLine = nodeStartLine + endLineOffset;
+					const startColumn =
+						(startLine === nodeStartLine ? nodeStartColumn : 1) +
+						startColumnOffset;
+					const endColumn =
+						(endLine === nodeStartLine ? nodeStartColumn : 1) +
+						endColumnOffset;
+
+					context.report({
+						loc: {
+							start: {
+								line: startLine,
+								column: startColumn,
+							},
+							end: {
+								line: endLine,
+								column: endColumn,
+							},
+						},
+						messageId: "unexpectedImportant",
+					});
+				}
+			},
+		};
+	},
+};

--- a/src/rules/no-important.js
+++ b/src/rules/no-important.js
@@ -1,5 +1,7 @@
 /**
- * @fileoverview Rule to disallow !important flags.
+ * @fileoverview Rule to disallow `!important` flags.
+ * @author thecalamiity
+ * @author Yann Bertrand
  */
 
 //-----------------------------------------------------------------------------
@@ -39,11 +41,12 @@ export default {
 	},
 
 	create(context) {
+		const importantPattern = /!(\s|\/\*.*?\*\/)*important/iu;
+
 		return {
 			Declaration(node) {
 				if (node.important) {
 					const declarationText = context.sourceCode.getText(node);
-					const importantPattern = /!(\s|\/\*.*?\*\/)*important/iu;
 					const importantMatch =
 						importantPattern.exec(declarationText);
 

--- a/src/util.js
+++ b/src/util.js
@@ -23,3 +23,28 @@
 export function isSyntaxMatchError(error) {
 	return typeof error.syntax === "string";
 }
+
+/**
+ * Finds the line and column offsets for a given offset in a string.
+ * @param {string} text The text to search.
+ * @param {number} offset The offset to find.
+ * @returns {{lineOffset:number,columnOffset:number}} The location of the offset.
+ */
+export function findOffsets(text, offset) {
+	let lineOffset = 0;
+	let columnOffset = 0;
+
+	for (let i = 0; i < offset; i++) {
+		if (text[i] === "\n") {
+			lineOffset++;
+			columnOffset = 0;
+		} else {
+			columnOffset++;
+		}
+	}
+
+	return {
+		lineOffset,
+		columnOffset,
+	};
+}

--- a/tests/rules/no-important.test.js
+++ b/tests/rules/no-important.test.js
@@ -1,0 +1,264 @@
+/**
+ * @fileoverview Tests for no-important rule.
+ */
+
+//------------------------------------------------------------------------------
+// Imports
+//------------------------------------------------------------------------------
+
+import rule from "../../src/rules/no-important.js";
+import css from "../../src/index.js";
+import { RuleTester } from "eslint";
+import dedent from "dedent";
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({
+	plugins: {
+		css,
+	},
+	language: "css/css",
+});
+
+ruleTester.run("no-important", rule, {
+	valid: [
+		"a { color: red; }",
+		"a { color: red; background-color: blue; }",
+		"a { color: red; transition: none; }",
+		"body { --custom-property: red; }",
+		"body { --custom-property: important; }",
+		"body { padding: 0; }",
+		"a { color: red; -moz-transition: bar }",
+		"@font-face { font-weight: 100 400 }",
+		'@property --foo { syntax: "*"; inherits: false; }',
+		"a { --my-color: red; color: var(--my-color) }",
+		{
+			code: "a { my-custom-color: red; }",
+			languageOptions: {
+				customSyntax: {
+					properties: {
+						"my-custom-color": "<color>",
+					},
+				},
+			},
+		},
+	],
+	invalid: [
+		{
+			code: "a { color: red !important; }",
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 1,
+					column: 16,
+					endLine: 1,
+					endColumn: 26,
+				},
+			],
+		},
+		{
+			code: "a { color:red!important; }",
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 1,
+					column: 14,
+					endLine: 1,
+					endColumn: 24,
+				},
+			],
+		},
+		{
+			code: "a { padding: 10px 20px 30px 40px ! important; }",
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 1,
+					column: 34,
+					endLine: 1,
+					endColumn: 45,
+				},
+			],
+		},
+		{
+			code: "a { border: 0 !IMPORTANT; }",
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 1,
+					column: 15,
+					endLine: 1,
+					endColumn: 25,
+				},
+			],
+		},
+		{
+			code: "a { color: red !important; margin: 0 ! important; }",
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 1,
+					column: 16,
+					endLine: 1,
+					endColumn: 26,
+				},
+				{
+					messageId: "unexpectedImportant",
+					line: 1,
+					column: 38,
+					endLine: 1,
+					endColumn: 49,
+				},
+			],
+		},
+		{
+			code: dedent`
+				a {
+					color: red
+						!important;
+				}
+			`,
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 3,
+					column: 3,
+					endLine: 3,
+					endColumn: 13,
+				},
+			],
+		},
+		{
+			code: dedent`
+				a {
+					color: red
+						! important;
+				}
+			`,
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 3,
+					column: 3,
+					endLine: 3,
+					endColumn: 14,
+				},
+			],
+		},
+		{
+			code: "a { color: red /* comment */ !important; }",
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 1,
+					column: 30,
+					endLine: 1,
+					endColumn: 40,
+				},
+			],
+		},
+		{
+			code: dedent`
+				a {
+					color: red /* comment */
+						!important;
+				}
+			`,
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 3,
+					column: 3,
+					endLine: 3,
+					endColumn: 13,
+				},
+			],
+		},
+		{
+			code: "a { color: red !/* comment */important; }",
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 1,
+					column: 16,
+					endLine: 1,
+					endColumn: 39,
+				},
+			],
+		},
+		{
+			code: "a { color: red ! /* comment */ important; }",
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 1,
+					column: 16,
+					endLine: 1,
+					endColumn: 41,
+				},
+			],
+		},
+		{
+			code: dedent`
+				a {
+					color: red
+						!/* comment */important;
+				}
+			`,
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 3,
+					column: 3,
+					endLine: 3,
+					endColumn: 26,
+				},
+			],
+		},
+		{
+			code: dedent`
+				a {
+					color: red
+						! /* comment */ important;
+				}
+			`,
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 3,
+					column: 3,
+					endLine: 3,
+					endColumn: 28,
+				},
+			],
+		},
+		{
+			code: dedent`
+				a {
+					color: red
+						!important;
+					margin: 0
+						! important;
+				}
+			`,
+			errors: [
+				{
+					messageId: "unexpectedImportant",
+					line: 3,
+					column: 3,
+					endLine: 3,
+					endColumn: 13,
+				},
+				{
+					messageId: "unexpectedImportant",
+					line: 5,
+					column: 3,
+					endLine: 5,
+					endColumn: 14,
+				},
+			],
+		},
+	],
+});

